scripts/pylib/twister/twisterlib/testplan.py

```diff
@@ -1015,11 +1021,94 @@ def apply_filters(self, **kwargs):
+        for instance in self.instances.values():
+            if not instance.testsuite.required_images:
+                continue
+            if instance.status:
+                # do not proceed if the test is already filtered
+                continue
+
+            if self.options.subset:
+                reason = "Required images are not supported with --subsets"
+                instance.add_filter(reason, Filters.REQUIRED_IMAGE)
+                logger.debug(f"{instance.name}: {reason}")
+                continue
+
+            if self.options.runtime_artifact_cleanup:
+                reason = "Required images are not supported with --runtime-artifact-cleanup"
+                instance.add_filter(reason, Filters.REQUIRED_IMAGE)
+                logger.debug(f"{instance.name}: {reason}")
+                continue
+
+            if instance.testsuite.no_own_image or not Path(
+                    Path(instance.testsuite.source_dir) / 'CMakeLists.txt').exists():
+                instance.no_own_image = True
+                # filter out qemu platforms, because for tests must use origin path to QEMU_PIPE,
+                # can be done, but first better to refactor QemuHandler
+                if instance.platform.simulation == 'qemu':
+                    instance.add_filter("QEMU with no own image not supported", Filters.REQUIRED_IMAGE)
+                    logger.debug(f"{instance.name}: QEMU with no own image not supported")
+                    continue
+
+                # check if platform is correct in required images
+                first_image = instance.testsuite.required_images[0]
+                req_platform, _ = self._get_required_platform_and_id(first_image, instance)
+                if req_platform != instance.platform.name:
+                    instance.add_filter(f"Wrong platform in required image: {first_image}", Filters.REQUIRED_IMAGE)
+                    logger.warning(f"{instance.name}: If there is no own image, the first required image"
+                                   " should be for the same platform as the test")
+                    continue
+
+            for required_image in instance.testsuite.required_images:
+                req_instance = self._find_required_instance(required_image, instance)
+
+                if not req_instance:
+                    instance.add_filter(f"Missing required image {required_image}", Filters.REQUIRED_IMAGE)
+                    logger.debug(f"{instance.name}: Required image '{required_image}' was not found."
+                                 " Please verify if required image is provided with --testsuite-root")
+                    break
+
+                if req_instance.status:
+                    # check if required image is filtered because is not runnable
+                    if self.options.device_testing and not req_instance.run and \
+                            len(req_instance.filters) == 1 and req_instance.reason == "Not runnable on device":
```

Comment:
The magic string `reason` comparison seems like a not great solution, but I do not see anything better than that currently.